File: server/src/pdf/extract.ts
import { extractText as unpdfExtractText, getDocumentProxy } from "unpdf";

export interface ExtractResult {
  text: string;
  numpages: number;
  info?: Record<string, unknown>;
}

export async function extractText(pdfBytes: Uint8Array | Buffer): Promise<ExtractResult> {
  const bytes = pdfBytes instanceof Uint8Array ? pdfBytes : new Uint8Array(pdfBytes);
  const pdf = await getDocumentProxy(bytes);
  const result = await unpdfExtractText(pdf, { mergePages: true });
  const text = typeof result.text === "string" ? result.text : (result.text as string[]).join("\n");
  return {
    text,
    numpages: pdf.numPages,
  };
}

export function highlightInjections(
  extractedText: string,
  payloads: Array<{ id: string; text: string }>,
): Array<{ id: string; start: number; end: number; excerpt: string }> {
  const hits: Array<{ id: string; start: number; end: number; excerpt: string }> = [];
  const hay = extractedText;
  const hayLower = hay.toLowerCase();

  for (const p of payloads) {
    const needle = p.text.slice(0, 80);
    const idx = hayLower.indexOf(needle.toLowerCase());
    if (idx >= 0) {
      hits.push({
        id: p.id,
        start: idx,
        end: idx + needle.length,
        excerpt: hay.slice(idx, idx + Math.min(120, p.text.length)),
      });
      continue;
    }
    const token = p.text.split(/\s+/).slice(0, 6).join(" ");
    const tIdx = hayLower.indexOf(token.toLowerCase());
    if (tIdx >= 0) {
      hits.push({
        id: p.id,
        start: tIdx,
        end: tIdx + token.length,
        excerpt: hay.slice(tIdx, tIdx + 120),
      });
    }
  }
  return hits;
}

File: server/app/pdf/extract.py
from __future__ import annotations

import io
from typing import Any

from pypdf import PdfReader


def extract_text(pdf_bytes: bytes) -> dict[str, Any]:
    reader = PdfReader(io.BytesIO(pdf_bytes))
    parts: list[str] = []
    for page in reader.pages:
        parts.append(page.extract_text() or "")
    return {
        "text": "\n".join(parts),
        "numpages": len(reader.pages),
    }


def highlight_injections(
    extracted_text: str,
    payloads: list[dict[str, Any]],
) -> list[dict[str, Any]]:
    hits: list[dict[str, Any]] = []
    hay = extracted_text
    hay_lower = hay.lower()

    for p in payloads:
        needle = p["text"][:80]
        idx = hay_lower.find(needle.lower())
        if idx >= 0:
            hits.append(
                {
                    "id": p["id"],
                    "start": idx,
                    "end": idx + len(needle),
                    "excerpt": hay[idx : idx + min(120, len(p["text"]))],
                }
            )
            continue
        token = " ".join(p["text"].split()[:6])
        t_idx = hay_lower.find(token.lower())
        if t_idx >= 0:
            hits.append(
                {
                    "id": p["id"],
                    "start": t_idx,
                    "end": t_idx + len(token),
                    "excerpt": hay[t_idx : t_idx + 120],
                }
            )
    return hits
